src/stirling/source_connectors/socket_tracer/uprobe_symaddrs.h

```diff
@@ -19,14 +19,37 @@
 #pragma once
 
 #include "src/common/base/base.h"
+#include "src/common/system/proc_parser.h"
 #include "src/stirling/obj_tools/dwarf_reader.h"
 #include "src/stirling/obj_tools/elf_reader.h"
 #include "src/stirling/source_connectors/socket_tracer/bcc_bpf_intf/symaddrs.h"
 #include "src/stirling/utils/detect_application.h"
 
+DECLARE_bool(openssl_force_raw_fptrs);
+
 namespace px {
 namespace stirling {
 
+using system::ProcParser;
+
+class RawFptrManager : NotCopyMoveable {
+
+  public:
+    explicit RawFptrManager(obj_tools::ElfReader* elf_reader, system::ProcParser* proc_parser, std::string lib_path);
+    ~RawFptrManager();
+
+    template <class T>
+    StatusOr<T*> RawSymbolToFptr(const std::string& symbol_name);
+    StatusOr<bool> Init();
+
+  private:
+    obj_tools::ElfReader* elf_reader_;
+    ProcParser* proc_parser_;
```

Comment:
Please let me know if how I'm handling these pointers is dangerous. I'm assuming that since the caller is using a `std::unique_ptr` that I won't have to assume any responsibility of managing its lifecycle. This is an area where I'm not confident what the correct thing to do is.

Reply:
Should be fine.

We have a few other places use ElfReader* ProcParser* in function as input argument to be used inside.

However, based on what I read, I think a cleaner way is to let RawFptrManager's ctor accept ProcessMap, and instead letter caller of RawFptrManager to call ProcParser to get the ProcessMap. That should not affect the overall flow of RawFptrManager.

WDYT?

Reply:
I actually considered this, but I implemented it this way for the following reasons.

When `dlsym` works, we won't perform any additional work (aside from constructing the object). Doing it as you described means we will parse `/proc/$PID/maps` and `dlopen` the library an extra time, regardless if we need to fall back on this implementation. We could defer the `ProcParser::GetExecutableMapEntry` call until we need it, but then the `ProcParser` will need to be passed much further in the `uprobe_symaddrs` code. So I thought this separation was cleaner in order to defer retrieving the ProcessMap.

In addition, if the RawFptrManager's ctor accepted a ProcessMap it means that the caller could construct an object with a mismatched `lib_path` and `process_map`. See the example below for what I mean

```
# ctr definition
RawFptrManager(obj_tools::ElfReader* elf_reader, ProcessMap process_map, std::string lib_path);

std::string libpath = " ... ";
auto map = proc_parser_->GetExecutableMapEntry(pid, "not the lib path above");

// dlopen will get called on an incorrect binary
RawFptrManager(elf_reader, map, libpath);
```

Let me know what you think on that justification.

Reply:
@yzhao1012 just checking in to see if you saw my previous comment.